### PR TITLE
First simple CPU bounded static thread pool

### PR DESCRIPTION
**Pull request by kononovk**

### CODE REVIEW

File: src/executor/thread_pool.cpp [hunk 1]
```diff
@@ -10,86 +10,131 @@ namespace yaclib::executor {
...
+      if (_state != Running) {
+        return;
+      }
+      _tasks.PushBack(task.release());
+    }
+    _cv.notify_one();
   }
 
   void Stop() final {
     // TODO(kononovk): #2 After call: сalling Execute() for all,
```
Comment (MBkkt) [inline src/executor/thread_pool.cpp:54]: Remove todo

File: src/executor/thread_pool.cpp [hunk 2]
```diff
@@ -10,86 +10,131 @@ namespace yaclib::executor {
...
   void Close() final {
     // TODO(kononovk): #2 After call: сalling Execute() for all, does nothing
+    ChangeState(Closed);
+    _cv.notify_all();
   }
 
   void Cancel() final {
     // TODO(kononovk): #2 After call: сalling Execute() for all,
     //  does nothing and remove all not
+    ChangeState(Stopped);
```
Comment (MBkkt) [inline src/executor/thread_pool.cpp:69]: Doesn't work

File: src/executor/thread_pool.cpp [hunk 3]
```diff
@@ -10,86 +10,131 @@ namespace yaclib::executor {
...
+    _cv.notify_all();
   }
 
   void Wait() final {
     // TODO(kononovk): #2 After call: this ThreadPool haven't running threads
+    // TODO: use lock or wait while thredpool initialized
+    while (!_threads.IsEmpty()) {
+      auto thread_ptr = _threads.PopBack();
+      thread_ptr->Join();
```
Comment (MBkkt) [inline src/executor/thread_pool.cpp:78]: Use thread_factory from other pr

File: src/executor/thread_pool.cpp [hunk 4]
```diff
@@ -10,86 +10,131 @@ namespace yaclib::executor {
...
     //  does nothing and remove all not
+    ChangeState(Stopped);
+    _cv.notify_all();
   }
 
   void Wait() final {
     // TODO(kononovk): #2 After call: this ThreadPool haven't running threads
+    // TODO: use lock or wait while thredpool initialized
+    while (!_threads.IsEmpty()) {
+      auto thread_ptr = _threads.PopBack();
```
Comment (MBkkt) [inline src/executor/thread_pool.cpp:77]: PopFront()

File: src/executor/thread_pool.cpp [hunk 5]
```diff
@@ -10,86 +10,131 @@ namespace yaclib::executor {
...
+        _cv.wait(lk);
+      }
+      if (_state == Stopped) {
+        return;
       }
-      auto current = _tasks.PopFront();
-      guard.unlock();
+      auto current = ITaskPtr{_tasks.PopFront()};
+      lk.unlock();
+      _cv.notify_one();
```
Comment (MBkkt) [inline src/executor/thread_pool.cpp:105]: WTF

File: test/thread_pool.cpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+namespace {
+
+using namespace yaclib;
+using namespace std::chrono_literals;
+
+GTEST_TEST(single_threaded, just_works) {
+  auto factory =
+      executor::MakeThreadFactory(0, std::thread::hardware_concurrency());
+  auto tp =
+      executor::MakeThreadPool(factory, std::thread::hardware_concurrency(),
```
Comment (MBkkt) [inline test/thread_pool.cpp:21]: Make different function as in another pr

File: test/thread_pool.cpp [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                               std::thread::hardware_concurrency());
+  bool ready{false};
+  tp->Execute(CreateTask([&ready] {
```
Comment (MBkkt) [inline test/thread_pool.cpp:24]: MakeTask

File: test/thread_pool.cpp [hunk 8]
```diff
@@ -0,0 +1,199 @@
+#include <cpu_time.hpp>
```
Comment (MBkkt) [inline test/thread_pool.cpp:1]: Why are tests so simple?

**Review by MBkkt — CHANGES_REQUESTED** [review]

File: src/executor/thread_pool.cpp [hunk 9]
```diff
@@ -10,86 +10,131 @@ namespace yaclib::executor {
...
-     */
+  }
+
+  void ChangeState(PoolState state) {
+    std::lock_guard lk{_m};
+    _state = state;
   }
 
   container::intrusive::List<IThread> _threads;
   container::intrusive::List<ITask> _tasks;
```
Comment (MBkkt) [inline src/executor/thread_pool.cpp:117]: Leak

File: src/executor/thread_pool.cpp [hunk 10]
```diff
@@ -10,86 +10,131 @@ namespace yaclib::executor {
...
+    Running,  // can add tasks and execute them
+    Closed,   // can not add new tasks, execute old tasks now
+    Stopped,  // stop execute current tasks from intrusive list
+  };
   void Loop() {
-    /* TODO(kononovk): #2 implement
+    // TODO(kononovk): #2 implement
     while (true) {
-      std::unique_lock guard{_m};
+      std::unique_lock lk{_m};
```
Comment (MBkkt) [inline src/executor/thread_pool.cpp:93]: guard good name btw

File: src/executor/thread_pool.cpp [hunk 11]
```diff
@@ -10,86 +10,131 @@ namespace yaclib::executor {
...
   ThreadPool(IThreadFactoryPtr factory, size_t min_threads_count,
              size_t max_threads_count)
-      : _min_threads_count{min_threads_count},
+      : _factory(factory),
+        _min_threads_count{min_threads_count},
         _max_threads_count{max_threads_count} {
     auto loop_functor = [this] {
+      current_thread_pool = this;
       Loop();
     };
```
Comment (MBkkt) [inline src/executor/thread_pool.cpp:26]: current_thread_pool = nullptr

File: src/executor/thread_pool.cpp [hunk 12]
```diff
@@ -10,86 +10,131 @@ namespace yaclib::executor {
 
 namespace {
 
-class ThreadPool final : public IThreadPool {
+thread_local IThreadPool* current_thread_pool;
```
Comment (MBkkt) [inline src/executor/thread_pool.cpp:13]: `tlCurrentThreadPool`/`tCurrentThreadPool`/`lCurrentThreadPool` thread local
`sCurrentThreadPool` static
`kCurrentThreadPool` constexpr/const static
`gCurrentThreadPool` global
What do you think?

File: src/executor/thread_pool.cpp [hunk 13]
```diff
@@ -10,86 +10,131 @@ namespace yaclib::executor {
...
+
+class ThreadPool final : public IThreadPool,
+                         std::enable_shared_from_this<ThreadPool> {
```
Comment (MBkkt) [inline src/executor/thread_pool.cpp:16]: Why?

File: include/yaclib/executor/thread_factory.hpp [hunk 14]
```diff
@@ -1,32 +1,39 @@
...
+IThreadFactoryPtr MakeThreadFactory(size_t cache_threads, size_t max_threads);
+
+IThreadFactoryPtr MakeThreadFactory(IThreadFactoryPtr base, std::string name);
+
+IThreadFactoryPtr MakeThreadFactory(IThreadFactoryPtr base, size_t priority);
+
+IThreadFactoryPtr MakeThreadFactory(IThreadFactoryPtr base, IFuncPtr acquire,
+                                    IFuncPtr release);
+
+IThreadFactoryPtr MakeThreadFactory();
```
Comment (MBkkt) [inline include/yaclib/executor/thread_factory.hpp:37]: NIT
this overload should be first

File: test/CMakeLists.txt [hunk 15]
```diff
@@ -1,27 +1,55 @@
...
   set(GTEST_BOTH_LIBRARIES gtest gtest_main)
-  if (ENABLE_LTO)
-    set(CMAKE_INTERPROCEDURAL_OPTIMIZATION ON)
-  else ()
-    set(CMAKE_INTERPROCEDURAL_OPTIMIZATION OFF)
-  endif ()
+endif ()
+
+# Google Benchmark
+set(BENCHMARK_ENABLE_TESTING NO)
```
Comment (MBkkt) [inline test/CMakeLists.txt:17]: - [ ] in `if (NOT BENCHMARK_FOUND)`

File: test/thread_pool.cpp [hunk 16]
```diff
@@ -0,0 +1,300 @@
+#include "util/cpu_time.hpp"
```
Comment (MBkkt) [inline test/thread_pool.cpp:1]: - [ ] Tests are not done

File: test/thread_factory.cpp [hunk 17]
```diff
@@ -100,33 +101,42 @@ GTEST_TEST(multi_threaded, simple) {
...
-    release_thread->Join();
-    test_threads->Release(std::move(release_thread));
+    while (!threads.IsEmpty()) {
+      auto release_thread = executor::IThreadPtr{threads.PopFront()};
+      test_threads->Release(std::move(release_thread));
+    }
   }
 }
 
+GTEST_TEST(decorator, priority) {
```
Comment (MBkkt) [inline test/thread_factory.cpp:135]: Add TODO

**Review by MBkkt — CHANGES_REQUESTED** [review]

File: test/thread_pool.cpp (hunk 8 above)
Comment (kononovk) [reply test/thread_pool.cpp]: I will create task with implementation better test coverage

**Review by MBkkt — APPROVED** [review]

**Pull request merged**